[Quant] Add int16 quantization support

terrychenism opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #72213

Summary:
Add dtype=int32 support for observer

Test Plan:
python3 test/test_quantization.py TestObserver.test_per_tensor_observers

Reviewers:

Subscribers:

Tasks:

Tags:

terrychenism committed:
[Quant] Add int16 quantization support

Summary:
Add dtype=int32 support for observer

Test Plan:
python3 test/test_quantization.py TestObserver.test_per_tensor_observers_qint32

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on test/quantization/core/test_workflow_module.py:
```diff
@@ -129,6 +129,58 @@ def test_per_tensor_observers(self, qdtype, qscheme, reduce_range):
+            self.assertEqual(myobs.max_val, loaded_obs.max_val)
+            self.assertEqual(myobs.calculate_qparams(), loaded_obs.calculate_qparams())
+
     @given(qdtype=st.sampled_from((torch.qint8, torch.quint8)),
```
can we just add a torch.qint32 here instead of copying the test case?

jerryzh168 commented on torch/ao/quantization/utils.py:
```diff
@@ -266,6 +271,9 @@ def calculate_qmin_qmax(quant_min: int, quant_max: int, has_customized_qrange: b
                 quant_min, quant_max = 0, 255
+        elif dtype == torch.qint32:
+            # default range for int16
+            quant_min, quant_max = -32768, 32767
```
quant_min and quant_max for qint32 should be larger I think

jerryzh168 commented on torch/ao/quantization/utils.py:
```diff
@@ -245,10 +245,15 @@ def calculate_qmin_qmax(quant_min: int, quant_max: int, has_customized_qrange: b
+            ), "quantization range should be positive and not exceed the maximum bit range (=256)."
+        elif dtype == torch.qint32:
+            assert (
+                0 < qrange_len <= 2**16
```
I think we should use 2**32 here since this is qint32

jerryzh168 commented on torch/ao/quantization/utils.py:
```diff
@@ -245,10 +245,15 @@ def calculate_qmin_qmax(quant_min: int, quant_max: int, has_customized_qrange: b
             )
```
should we get `initial_quant_min` and `initial_quant_max` based on dtype? currently it is hardcoded to 0, 255

terrychenism committed:
Update on "[Quant] Add int16 quantization support"


Summary:
Add dtype=int32 support for observer

Test Plan:
python3 test/test_quantization.py TestObserver.test_per_tensor_observers_qint32

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

terrychenism committed:
Update on "[Quant] Add int16 quantization support"


Summary:
Add dtype=int32 support for observer

Test Plan:
python3 test/test_quantization.py TestObserver.test_per_tensor_observers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

terrychenism committed:
Update on "[Quant] Add int16 quantization support"


Summary:
Add dtype=int32 support for observer

Test Plan:
python3 test/test_quantization.py TestObserver.test_per_tensor_observers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

terrychenism committed:
Update on "[Quant] Add int16 quantization support"


Summary:
Add dtype=int32 support for observer

Test Plan:
python3 test/test_quantization.py TestObserver.test_per_tensor_observers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

terrychenism committed:
Update on "[Quant] Add int16 quantization support"


Summary:
Add dtype=int32 support for observer

Test Plan:
python3 test/test_quantization.py TestObserver.test_per_tensor_observers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]
